### PR TITLE
Re-introduce training controller for workload log training

### PR DESCRIPTION
**Pull request by AmartC**
This PR re-introduces the training controller for workload logs. It will also be communicating with the model training gateway plugin for training a model, getting the last workload parameters as well as the current status of the workload model.

### CODE REVIEW

File: training_controller/prepare_training_logs.py [hunk 1]
```diff
@@ -176,6 +176,57 @@ def fetch_training_logs_from_elasticsearch(
...
+            current_command = esdump_sample_command[:]
+            current_command[2] = current_command[2].format(start_ts, end_ts)
+            current_command[6] = current_command[6].format(
+                timestamps_esdump_num_logs_fetched[idx]
+            )
+            current_command[10] = current_command[10].format(
+                os.path.join(self.ES_DUMP_DIR, f"{filename}")
+            )
+            query_queue.append(current_command)
+        """
```
Comment (tybalex) [inline training_controller/prepare_training_logs.py]: doc strings should be placed at the first line of the functions.

File: Dockerfile [hunk 2]
```diff
@@ -1,8 +1,10 @@
 FROM rancher/opni-python-base:3.8-node14-elasticdump
 ENV NODE_TLS_REJECT_UNAUTHORIZED 0
+EXPOSE 80
```
Comment (tybalex) [inline Dockerfile]: why do we need to expose port 80 though?

File: training_controller/prepare_training_logs.py (hunk 1 above)
Comment (AmartC) [reply training_controller/prepare_training_logs.py]: That's not a doc string as much a multi-line comment discussing the last edge case so it's okay to put it here to give it context.

Comment (tybalex) [reply training_controller/prepare_training_logs.py]: if it's not a doc string , you should use `#`s to comment them..

File: training_controller/main.py [hunk 3]
```diff
@@ -4,134 +4,141 @@
...
+        workload_parameters_dict = json.loads(workload_parameters)
+        current_ts = int(time.time() * 1000)
+        await es_instance.index(
+            index="model-training-parameters",
+            body={"time": current_ts, "parameters": workload_parameters},
+        )
+        max_logs_for_training = PrepareTrainingLogs().get_num_logs_for_training()
+        logging.info(max_logs_for_training)
+        end_ts = int(time.time() * 1000)
+        start_ts = end_ts - 3600000
```
Comment (tybalex) [inline training_controller/main.py:145]: let's make magic number like this a CONSTANT in the next v0.6.1 PR

**Review by tybalex — DISMISSED** [review]

**Review by tybalex — DISMISSED** [review]

File: training_controller/main.py [hunk 4]
```diff
@@ -4,134 +4,205 @@
...
-        }
-    ]
-    logging.info(f"ES job {request_id} status update : {job_status}")
+def get_gpu_status():
     try:
-        async for ok, result in async_streaming_bulk(es, docs_to_update):
-            action, result = result.popitem()
-            if not ok:
-                logging.error(f"failed to index documents {docs_to_update}")
+        results = requests.get("http://opni-internal:11080/ModelTraining/gpu_info")
```
Comment (tybalex) [inline training_controller/main.py]: can you make the url a constant

File: training_controller/main.py [hunk 5]
```diff
@@ -4,134 +4,205 @@
...
-                        "payload": job_payload,
-                    }
-                    await update_es_job_status(
-                        request_id=hit["_id"], job_status="training scheduled"
-                    )
-                    await schedule_training_job(training_job_payload)
-        except (exceptions.NotFoundError, exceptions.TransportError) as e:
-            logging.error(e)
-        await asyncio.sleep(60)
+        response = await nw.request("gpu_service_running", b"check-up", timeout=1)
```
Comment (tybalex) [inline training_controller/main.py]: what happens If the GPU service is in the middle of a training job?

Comment (AmartC) [reply training_controller/main.py]: it returns unavailable then. This case is more for if the GPU is removed and the pod is pending but it'll also return unavailable in addition to the gpu_training_request variable being set to 1.

Comment (tybalex) [reply training_controller/main.py]: cool. as long as its an expected behavior then all good.

File: training_controller/main.py [hunk 6]
```diff
@@ -71,36 +72,42 @@ async def get_gpu_service_status():
...
     updated_workload_parameters = json.loads(workload_parameters)
-    current_ts = int(time.time() * 1000)
-    await es_instance.index(
-        index="model-training-parameters",
-        body={"time": current_ts, "parameters": workload_parameters},
-    )
-    model_training_necessary, workload_parameters_dict = check_training_necessary(
-        updated_workload_parameters
-    )
+    update_opensearch(workload_parameters)
```
Comment (tybalex) [inline training_controller/main.py]: await?

**Review by tybalex — DISMISSED** [review]
Look good for now. Refactor needed in the future: https://zube.io/rancher/opni/c/1319

**Review by tybalex — APPROVED** [review]

**Pull request merged**